# Mod3_Final Assessment - Data Exploration, Cleaning and Statistical Analysis

In this assessment, I work with two datasets related to an airline loyalty program. The goal is to explore their structure, identify data quality issues such as missing values or inconsistencies, and prepare the information for further statistical analysis and visualization.

The first dataset contains monthly flight activity, including flights booked, distance flown, and loyalty points. The second dataset contains customer profile information, such as location, education, salary, marital status, and loyalty program details.

As a first step, I perform an initial exploration of both datasets to understand their structure, review data types, detect null values and duplicates, and generate descriptive statistics. Then, I merge both datasets using the customer identifier in order to create a complete dataset for the next phases of the assessment.

After the exploration phase, a data cleaning process is carried out to ensure the quality and consistency of the dataset. This includes handling missing values in key variables, verifying the coherence of the data, and performing necessary transformations such as adjusting data types or correcting inconsistencies. These steps ensure that the dataset is properly prepared for statistical analysis.

## 1. Initial Exploration

In [4]:
# Import libraries

import pandas as pd
import numpy as np

# Display all columns when printing DataFrames
pd.set_option('display.max_columns', None) 

In [5]:
# Load the first dataset: Customer Flight Activity

df1 = pd.read_csv("../files/raw/Customer Flight Activity.csv")
df1.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0
2,100140,2017,1,6,0,6,1200,120.0,0,0
3,100214,2017,1,0,0,0,0,0.0,0,0
4,100272,2017,1,0,0,0,0,0.0,0,0


In [6]:
# Check the size of the first dataset

print(f"Customer Flight Activity contains {df1.shape[0]} rows and {df1.shape[1]} columns.")

Customer Flight Activity contains 405624 rows and 10 columns.


In [7]:
# Get general information about the first dataset

df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


In [8]:
# Analyze each column in the first dataset

print("COLUMN ANALYSIS\n")

for column in df1.columns:
    print("Column: ", column)
    print("Data type: ", df1[column].dtype)
    print(f"Null percentage: {round(df1[column].isna().mean() * 100, 2)}%")
    print("Unique values: ", df1[column].unique())
    print("Number of unique values: ", df1[column].nunique())
    print("----------------------\n")

COLUMN ANALYSIS

Column:  Loyalty Number
Data type:  int64
Null percentage: 0.0%
Unique values:  [100018 100102 100140 ... 999731 999788 999891]
Number of unique values:  16737
----------------------

Column:  Year
Data type:  int64
Null percentage: 0.0%
Unique values:  [2017 2018]
Number of unique values:  2
----------------------

Column:  Month
Data type:  int64
Null percentage: 0.0%
Unique values:  [ 1  9  2  3 11  4  5  7  6  8 10 12]
Number of unique values:  12
----------------------

Column:  Flights Booked
Data type:  int64
Null percentage: 0.0%
Unique values:  [ 3 10  6  0  8 11  9  4  7  5  2  1 12 13 14 16 15 17 18 19 20 21]
Number of unique values:  22
----------------------

Column:  Flights with Companions
Data type:  int64
Null percentage: 0.0%
Unique values:  [ 0  4  7  1  6  3  5  2 10  8  9 11]
Number of unique values:  12
----------------------

Column:  Total Flights
Data type:  int64
Null percentage: 0.0%
Unique values:  [ 3 14  6  0 15 11 12 10  8  9  7  5 16  2 

After this initial exploration of the Customer Flight Activity dataset, several points can be highlighted:

- There are no missing values in this dataset.
- All variables are numeric.
- None of the columns are constant, since all of them contain more than one unique value.
- Most columns are stored as integers, while **Points Accumulated** is stored as a float. This may be acceptable depending on how points are calculated, although it should be reviewed during the cleaning phase.

In [9]:
# Check for duplicated rows in the first dataset

print(f"Customer Flight Activity contains {df1.duplicated().sum()} duplicated entries.")
df1[df1.duplicated()]

Customer Flight Activity contains 1864 duplicated entries.


,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
42,101902,2017,1,0,0,0,0,0.0,0,0
227,112142,2017,1,0,0,0,0,0.0,0,0
478,126100,2017,1,0,0,0,0,0.0,0,0
567,130331,2017,1,0,0,0,0,0.0,0,0
660,135421,2017,1,0,0,0,0,0.0,0,0
...,...,...,...,...,...,...,...,...,...,...
404668,949628,2018,12,0,0,0,0,0.0,0,0
404884,960050,2018,12,0,0,0,0,0.0,0,0
405111,971370,2018,12,0,0,0,0,0.0,0,0
405410,988392,2018,12,0,0,0,0,0.0,0,0


In [10]:
# Descriptive statistics for the (numeric) variables in the first dataset

df1.describe().T

,count,mean,std,min,25%,50%,75%,max
Loyalty Number,405624.0,550037.873084,258935.286969,100018.0,326961.00,550834.0,772194.00,999986.0
Year,405624.0,2017.500000,0.500001,2017.0,2017.00,2017.5,2018.00,2018.0
Month,405624.0,6.500000,3.452057,1.0,3.75,6.5,9.25,12.0
Flights Booked,405624.0,4.115052,5.225518,0.0,0.00,1.0,8.00,21.0
Flights with Companions,405624.0,1.031805,2.076869,0.0,0.00,0.0,1.00,11.0
Total Flights,405624.0,5.146858,6.521227,0.0,0.00,1.0,10.00,32.0
Distance,405624.0,1208.880059,1433.155320,0.0,0.00,488.0,2336.00,6293.0
Points Accumulated,405624.0,123.692721,146.599831,0.0,0.00,50.0,239.00,676.5
Points Redeemed,405624.0,30.696872,125.486049,0.0,0.00,0.0,0.00,876.0
Dollar Cost Points Redeemed,405624.0,2.484503,10.150038,0.0,0.00,0.0,0.00,71.0


The dataset shows a highly skewed distribution across most variables, with many customers showing low activity and a smaller group reaching much higher values.

**Flights Booked** has a low median (1) compared with its mean (around 4), which suggests that most customers book few flights, while a small group books much more frequently. A similar pattern can be seen in **Distance** and **Points Accumulated**, where dispersion is high and some extreme values appear.

In contrast, **Points Redeemed** and **Dollar Cost Points Redeemed** are mostly concentrated around zero, suggesting that many customers do not frequently redeem their points.

Overall, the data suggests that customer activity is unevenly distributed, with a minority of highly active customers generating a large share of the activity.

In [11]:
# Load the second dataset: Customer Loyalty History

df2 = pd.read_csv("../files/raw/Customer Loyalty History.csv")
df2.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN


In [12]:
# Check the size of the second dataset

print(f"Customer Loyalty History contains {df2.shape[0]} rows and {df2.shape[1]} columns.")

Customer Loyalty History contains 16737 rows and 16 columns.


In [13]:
# Get general information about the second dataset

df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  str    
 2   Province            16737 non-null  str    
 3   City                16737 non-null  str    
 4   Postal Code         16737 non-null  str    
 5   Gender              16737 non-null  str    
 6   Education           16737 non-null  str    
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  str    
 9   Loyalty Card        16737 non-null  str    
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  str    
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: float64(4), 

In [14]:
# Analyze each column in the second dataset

print("COLUMN ANALYSIS\n")

for column in df2.columns:
    print("Column: ", column)
    print("Data type: ", df2[column].dtype)
    print(f"Null percentage: {round(df2[column].isna().mean() * 100, 2)}%")
    print("Unique values: ", df2[column].unique())
    print("Number of unique values: ", df2[column].nunique())
    print("----------------------\n")

COLUMN ANALYSIS

Column:  Loyalty Number
Data type:  int64
Null percentage: 0.0%
Unique values:  [480934 549612 429460 ... 776187 906428 652627]
Number of unique values:  16737
----------------------

Column:  Country
Data type:  str
Null percentage: 0.0%
Unique values:  <StringArray>
['Canada']
Length: 1, dtype: str
Number of unique values:  1
----------------------

Column:  Province
Data type:  str
Null percentage: 0.0%
Unique values:  <StringArray>
[             'Ontario',              'Alberta',     'British Columbia',
               'Quebec',                'Yukon',        'New Brunswick',
             'Manitoba',          'Nova Scotia',         'Saskatchewan',
         'Newfoundland', 'Prince Edward Island']
Length: 11, dtype: str
Number of unique values:  11
----------------------

Column:  City
Data type:  str
Null percentage: 0.0%
Unique values:  <StringArray>
[       'Toronto',       'Edmonton',      'Vancouver',           'Hull',
     'Whitehorse',        'Trenton',       '

After this initial exploration of the Customer Loyalty History dataset, the following points can be highlighted:

- This dataset contains missing values.
    - Some variables, such as **Cancellation Year**, **Cancellation Month**, and **Salary**, will need special attention during the cleaning phase due to the presence of null values.
- **Country** is a constant variable, since all records correspond to Canada.
- The dataset includes both numeric and categorical variables, which will require different types of analysis.

In [15]:
# Calculate the percentage of null values in the second dataset

nulls = round((df2.isna().sum() / df2.shape[0]) * 100, 2)
nulls[nulls > 0].sort_values(ascending=False)

Cancellation Year     87.65
Cancellation Month    87.65
Salary                25.32
dtype: float64

**Cancellation Year** and **Cancellation Month** have a high percentage of missing values (87.65%), which likely indicates that those customers are still active in the loyalty program and have not cancelled their membership.

**Salary** also shows a relevant proportion of missing values (25.32%), so this variable may require further treatment before analysis.

In [16]:
# Check for duplicated rows in the second dataset

print(f"Customer Loyalty History contains {df2.duplicated().sum()} duplicated entries.")

Customer Loyalty History contains 0 duplicated entries.


In [17]:
# Check whether Loyalty Number is unique in the second dataset

df2["Loyalty Number"].duplicated().sum()

np.int64(0)

No duplicated customer IDs were found in **Loyalty Number**, which suggests that each row represents a unique customer profile in this dataset.

In [18]:
# Descriptive statistics for the numeric variables in the second dataset

df2.describe().T

,count,mean,std,min,25%,50%,75%,max
Loyalty Number,16737.0,549735.880445,258912.132453,100018.00,326603.00,550434.00,772019.00,999986.00
Salary,12499.0,79245.609409,35008.297285,-58486.00,59246.50,73455.00,88517.50,407228.00
CLV,16737.0,7988.896536,6860.982280,1898.01,3980.84,5780.18,8940.58,83325.38
Enrollment Year,16737.0,2015.253211,1.979111,2012.00,2014.00,2015.00,2017.00,2018.00
Enrollment Month,16737.0,6.669116,3.398958,1.00,4.00,7.00,10.00,12.00
Cancellation Year,2067.0,2016.503145,1.380743,2013.00,2016.00,2017.00,2018.00,2018.00
Cancellation Month,2067.0,6.962748,3.455297,1.00,4.00,7.00,10.00,12.00


The numeric variables in this dataset show considerable variability, especially in **Salary** and **CLV**.

**Salary** presents a wide range of values, including some negative values, which may indicate data quality issues and should be reviewed during the cleaning phase. **CLV** also shows substantial dispersion, suggesting notable differences in customer value across the dataset.

**Enrollment Year**, **Enrollment Month**, **Cancellation Year**, and **Cancellation Month** mainly serve as temporal reference variables and will be more useful for contextual or segmented analysis than for direct numerical interpretation.

Overall, the numeric profile dataset includes some potentially useful business variables, but it also shows issues that should be addressed before deeper analysis.

In [19]:
# Descriptive statistics for the categorical variables in the second dataset

df2.describe(include="str").T

,count,unique,top,freq
Country,16737,1,Canada,16737
Province,16737,11,Ontario,5404
City,16737,29,Toronto,3351
Postal Code,16737,55,V6E 3D9,911
Gender,16737,2,Female,8410
Education,16737,5,Bachelor,10475
Marital Status,16737,3,Married,9735
Loyalty Card,16737,3,Star,7637
Enrollment Type,16737,2,Standard,15766


The dataset shows a fairly concentrated customer profile, with limited variability in some categorical variables.

All customers are from **Canada**, while **Ontario** and **Toronto** are the most frequent province and city, which suggests a geographic concentration. Gender is relatively balanced, although **Female** appears slightly more often.

Most customers have a **Bachelor's degree** and are **Married**, which indicates a fairly homogeneous demographic profile. In terms of the loyalty program, **Star** is the most common card type, and **Standard** is by far the most frequent enrollment type.

Overall, the customer base appears relatively similar in demographic and loyalty program characteristics, with clear concentration in some categories.

In [20]:
# We merge both datasets using "Loyalty Number" as the key, since it uniquely identifies each customer in both tables. This allows us to combine flight activity with customer profile information in a consistent and reliable way.

df = df1.merge(df2, on = "Loyalty Number")
df.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,100018,2017,1,3,0,3,1521,152.0,0,0,Canada,Alberta,Edmonton,T9G 1W3,Female,Bachelor,92552.0,Married,Aurora,7919.20,Standard,2016,8,NaN,NaN
1,100102,2017,1,10,4,14,2030,203.0,0,0,Canada,Ontario,Toronto,M1R 4K3,Male,College,NaN,Single,Nova,2887.74,Standard,2013,3,NaN,NaN
2,100140,2017,1,6,0,6,1200,120.0,0,0,Canada,British Columbia,Dawson Creek,U5I 4F1,Female,College,NaN,Divorced,Nova,2838.07,Standard,2016,7,NaN,NaN
3,100214,2017,1,0,0,0,0,0.0,0,0,Canada,British Columbia,Vancouver,V5R 1W3,Male,Bachelor,63253.0,Married,Star,4170.57,Standard,2015,8,NaN,NaN
4,100272,2017,1,0,0,0,0,0.0,0,0,Canada,Ontario,Toronto,P1L 8X8,Female,Bachelor,91163.0,Divorced,Star,6622.05,Standard,2014,1,NaN,NaN


In [21]:
# Check the size of the merged dataset

print(f"Our final dataset contains {df.shape[0]} rows and {df.shape[1]} columns.")

Our final dataset contains 405624 rows and 25 columns.


In [22]:
# Check null values in the merged dataset

df_nulls = round((df.isna().sum() / df.shape[0]) * 100, 2)
df_nulls[df_nulls > 0].sort_values(ascending=False)

Cancellation Year     87.66
Cancellation Month    87.66
Salary                25.31
dtype: float64

After merging both datasets, the missing values observed in the customer profile variables remain consistent with the issues previously identified in the loyalty history dataset. No new structural problems appear to have been introduced by the merge.

In [23]:
# Before starting the cleaning phase, the merged dataset is saved to ensure reproducibility and to preserve the original combined structure of the data.

df.to_csv("../files/processed/merged_dataset_v1.csv", index=False)

## 2. Data Cleaning

In [24]:
# First, create a copy of our merged dataset for safety

df_copy = df.copy()

df_copy.head(2)

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,100018,2017,1,3,0,3,1521,152.0,0,0,Canada,Alberta,Edmonton,T9G 1W3,Female,Bachelor,92552.0,Married,Aurora,7919.20,Standard,2016,8,NaN,NaN
1,100102,2017,1,10,4,14,2030,203.0,0,0,Canada,Ontario,Toronto,M1R 4K3,Male,College,NaN,Single,Nova,2887.74,Standard,2013,3,NaN,NaN


We will start deleting or tratando our missing values.
If we check about next steps of this proyect, we don't need to use Cancellation Year or Cancellation Month for any visualization. So, we are going to check quickly if our hipothesis is right and there are nulls because they are still clients. If that, we will replace NaN for 0, to keep the data type.

Cancellation Year and Cancellation Month contain a high percentage of null values. These missing values do not indicate data quality issues, but rather that the customer has not cancelled their membership.

Since no further analysis involving cancellation dates is required, these null values are preserved as they carry meaningful information and do not need to be modified.

In [25]:
# Salary will be needed for further visualizations. We are going to analyze the missing values

df_copy["Salary"].describe()

count    302952.000000
mean      79268.825953
std       34992.133508
min      -58486.000000
25%       59262.000000
50%       73479.000000
75%       88612.000000
max      407228.000000
Name: Salary, dtype: float64

Salary contains missing values and some invalid negative values, which are not realistic for this variable.

In [26]:
# Analyzing negative values

df_copy[df_copy["Salary"] < 0] 

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
293,115505,2017,1,0,0,0,0,0.0,0,0,Canada,Newfoundland,St. John's,A1C 6H9,Male,Bachelor,-10605.0,Married,Nova,5860.17,2018 Promotion,2018,4,NaN,NaN
961,152016,2017,1,0,0,0,0,0.0,0,0,Canada,Ontario,Toronto,P1J 8T7,Female,Bachelor,-58486.0,Married,Aurora,5067.21,2018 Promotion,2018,2,NaN,NaN
1740,194065,2017,1,0,0,0,0,0.0,0,0,Canada,Ontario,Sudbury,M5V 1G5,Female,Bachelor,-31911.0,Married,Nova,2888.85,2018 Promotion,2018,2,NaN,NaN
2111,212128,2017,1,3,0,3,1959,195.0,0,0,Canada,Ontario,Toronto,P2T 6G3,Male,Bachelor,-49001.0,Married,Nova,3130.68,2018 Promotion,2018,2,NaN,NaN
2464,232755,2017,1,0,0,0,0,0.0,0,0,Canada,British Columbia,Vancouver,V1E 4R6,Female,Bachelor,-46683.0,Single,Nova,4787.81,2018 Promotion,2018,3,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
400664,734647,2018,12,15,0,15,1440,144.0,0,0,Canada,Saskatchewan,Regina,S1J 3C5,Male,Bachelor,-46303.0,Married,Nova,11280.73,2018 Promotion,2018,4,NaN,NaN
401739,790475,2018,12,7,1,8,1056,105.0,0,0,Canada,Ontario,Trenton,K8V 4B2,Female,Bachelor,-34079.0,Married,Nova,12913.50,2018 Promotion,2018,2,NaN,NaN
403876,906428,2018,12,13,0,13,2912,291.0,0,0,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN
404318,430398,2018,12,5,5,10,660,66.0,539,44,Canada,Newfoundland,St. John's,A1C 6H9,Male,Bachelor,-17534.0,Married,Nova,49423.80,2018 Promotion,2018,3,NaN,NaN


480 entries contains negative values on Salary. As mentioned, negative salary values are considered data errors, so they are treated as missing values. Instead of modifying them arbitrarily, they are replaced with NaN and later imputed using a more reliable approach.

In [27]:
df_copy.loc[df_copy["Salary"] < 0, "Salary"] = np.nan

In [28]:
# Missing values are imputed using the median salary within each education group
# This approach allows for a more realistic estimation, as salary is likely to vary depending on the level of education, and the median is robust to outliers
# Median is preferred over the mean to reduce the impact of extreme values

df_copy["Salary"] = df_copy.groupby("Education")["Salary"].transform(lambda x: x.fillna(x.median()))